File: javascript/src/agent/covfixer/triage.ts
/**
 * Triage: normalize an arbitrary coverage report into source files with meaningful
 * uncovered logic (LLM-backed).
 */
import type { BaseLlm } from '@google/adk';

import { type FileWork, extractJsonArray } from '../fixflow/index';
import { generateText } from '../setup/generate';
import { prompts } from './loader';

/** Use the LLM to find the source files with meaningful uncovered logic. */
export async function triage(llm: BaseLlm, report: string): Promise<FileWork[]> {
  const out = await generateText(llm, prompts.mustGet('triage'), report);
  const work = parseTriage(out);
  if (work.length === 0) {
    throw new Error('triage: no meaningful uncovered files found in report');
  }
  return work;
}

export function parseTriage(out: string): FileWork[] {
  const js = extractJsonArray(out);
  if (js === '') {
    throw new Error('triage: no JSON array in model output');
  }
  let files: unknown;
  try {
    files = JSON.parse(js);
  } catch (err) {
    throw new Error(`triage: decode triage JSON: ${(err as Error).message}`);
  }
  const work: FileWork[] = [];
  if (Array.isArray(files)) {
    for (const f of files) {
      const path = f && typeof f === 'object' && typeof f.path === 'string' ? f.path : '';
      if (path.trim() !== '') {
        const uncovered = Array.isArray(f.uncovered) ? f.uncovered.map(String) : [];
        work.push({ path, items: uncovered });
      }
    }
  }
  return work;
}

File: python/automation_agent/agent/covfixer/triage.py
"""Triage: normalize an arbitrary coverage report into source files with meaningful
uncovered logic (LLM-backed).
"""

from __future__ import annotations

import json

from google.adk.models import BaseLlm

from automation_agent.agent import setup
from automation_agent.agent.covfixer.loader import prompts
from automation_agent.agent.fixflow import FileWork, extract_json_array


async def triage(llm: BaseLlm, report: str) -> list[FileWork]:
    """Use the LLM to normalize a coverage report into the source files with meaningful
    uncovered logic."""
    out = await setup.generate_text(llm, prompts.must_get("triage"), report)
    work = _parse_triage(out)
    if not work:
        raise ValueError("triage: no meaningful uncovered files found in report")
    return work


def _parse_triage(out: str) -> list[FileWork]:
    js = extract_json_array(out)
    if js == "":
        raise ValueError("triage: no JSON array in model output")
    try:
        files = json.loads(js)
    except (ValueError, TypeError) as exc:
        raise ValueError(f"triage: decode triage JSON: {exc}") from exc
    work: list[FileWork] = []
    for f in files:
        path = (f.get("path") or "") if isinstance(f, dict) else ""
        if isinstance(path, str) and path.strip() != "":
            uncovered = f.get("uncovered") or []
            work.append(FileWork(path=path, items=list(uncovered)))
    return work
